src/gpuarray_util.c

```diff
@@ -130,3 +130,45 @@ int gpuarray_type_flags(int init, ...) {
   va_end(ap);
   return flags;
 }
+
+static inline void shiftdown(ssize_t *base, unsigned int i, unsigned int nd) {
+  if (base != NULL)
+    memmove(&base[i], &base[i+1], (nd - i - 1)*sizeof(size_t));
+}
+
+void gpuarray_elemwise_collapse(unsigned int n, unsigned int *_nd,
```

Comment:
The python version does argument validity checking, this only does dimension collapsing and you should have set the relevant broadcastable dimensions strides to 0 beforehand.

I'm working on the theano part, but it's trickier than I thought.